cpp/dolfinx/io/ADIOS2Writers.h

```diff
@@ -830,6 +830,333 @@ void vtx_write_mesh_from_space(adios2::IO& io, adios2::Engine& engine,
+    std::array<std::int64_t, 2> local_range
+        = geometry.index_map()->local_range();
+
+    std::span<const T> x = geometry.x();
+    std::size_t gdim = geometry.dim();
+
+    // Trim geometry to gdim and trunctate for owned entities only
+    std::vector<T> buffer(num_xdofs_local * gdim);
+    for (std::size_t i = 0; i < num_xdofs_local; ++i)
+      for (std::size_t j = 0; j < gdim; ++j)
+        buffer[i * gdim + j] = x[3 * i + j];
+
+    // Define geometry variable and "put"
+    adios2::Variable<T> local_geometry = impl_adios2::define_variable<T>(
+        io, "geometry", {num_xdofs_global, gdim}, {local_range[0], 0},
+        {num_xdofs_local, gdim});
+    engine.Put<T>(local_geometry, buffer.data());
+  }
+
+  // Extract topology (cell) information
+  auto topology = mesh.topology();
+  assert(topology);
+  auto cell_imap = topology->index_map(topology->dim());
+  const std::size_t num_cells_local = cell_imap->size_local();
+  const std::size_t num_cells_global = cell_imap->size_global();
+  const std::size_t local_range_cell = cell_imap->local_range()[0];
+
+  // Extract geometry information
+  if (geometry.cmaps().size() > 1)
+    throw std::runtime_error("Multiple coordinate maps not supported");
+  fem::ElementDofLayout geom_layout = geometry.cmaps()[0].create_dof_layout();
+  std::size_t num_dofs_per_cell
+      = geom_layout.num_entity_closure_dofs(geometry.dim());
+  auto geometry_imap = geometry.index_map();
+  auto geometry_dmap = geometry.dofmap();
+
+  // Convert local geometry to global geometry
+  std::vector<std::int64_t> connectivity_out(num_cells_local
+                                             * num_dofs_per_cell);
+  std::span<const std::int32_t> local_dmap(geometry_dmap.data_handle(),
+                                           num_cells_local * num_dofs_per_cell);
+  geometry_imap->local_to_global(local_dmap, connectivity_out);
+
+  // Put local topology in global dataset
+  adios2::Variable<std::int64_t> local_topology
+      = impl_adios2::define_variable<std::int64_t>(
+          io, "connectivity", {num_cells_global, num_dofs_per_cell},
+          {local_range_cell, 0}, {num_cells_local, num_dofs_per_cell});
+  engine.Put<std::int64_t>(local_topology, connectivity_out.data());
+
+  // Compute mesh permutations and store them to file. These are only
+  // needed for when we want to checkpoint functions.
+  mesh.topology_mutable()->create_entity_permutations();
+  const std::vector<std::uint32_t>& cell_perm
+      = topology->get_cell_permutation_info();
+  adios2::Variable<std::uint32_t> local_cell_perm
+      = impl_adios2::define_variable<std::uint32_t>(
+          io, "CellPermutations", {num_cells_global}, {local_range_cell},
+          {num_cells_local});
+  engine.Put<std::uint32_t>(local_cell_perm, cell_perm.data());
+
+  engine.PerformPuts();
+}
+
+} // namespace impl_checkpoint
+
+/// Read and write mesh::Mesh objects with ADIOS2
+template <std::floating_point T>
+class Checkpointer
```

Comment:
It doesn't seem logical that this class write and read - beyond the ADIOS2 handles the member data and methods are not shared across read and write. And the `ADIOS2Writer` class seems to already handle the ADISO2 handles.